tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -12,90 +12,457 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  TargetFunctionInfo(const string callee_name_, bool check_signature_,
+                     absl::Span<const PrimitiveType> input_types_,
+                     const PrimitiveType output_type_)
+      : callee_name(callee_name_),
+        check_signature(check_signature_),
+        input_types(input_types_),
+        result_type(output_type_) {}
+  // Potential signature mistamatch - use bitcast if specified by the target.
+  TargetFunctionInfo(const string callee_name_, bool check_signature_,
+                     absl::Span<const PrimitiveType> input_types_,
+                     const PrimitiveType output_type_, bool use_bitcast_)
+      : callee_name(callee_name_),
+        check_signature(check_signature_),
+        input_types(input_types_),
+        result_type(output_type_),
+        use_bitcast(use_bitcast_) {}
+  // Device function name.
+  const string callee_name;
+  // Check signature of the device function and if needed generate casts.
+  bool check_signature;
+  // Inpute types accespted by the device function.
+  absl::Span<const PrimitiveType> input_types;
+  // Result type of the device function.
+  PrimitiveType result_type;
+  // Use bitcast to generate casts if the desired signature at the call site
+  // does not match the signature of the device function.
+  absl::optional<bool> use_bitcast;
+};
+
+// Wrapper structure to carry either information about the intrinsic
+// or device function for NVPTX/AMDGPU.
+// Note: Only intrinsic or device function - not both - is expected.
+struct TargetInfo {
+  absl::optional<struct TargetIntrinsicInfo> target_intrinsic_info;
+  absl::optional<struct TargetFunctionInfo> target_function_info;
+  TargetInfo(struct TargetIntrinsicInfo x) : target_intrinsic_info(x) {}
+  TargetInfo(struct TargetFunctionInfo y) : target_function_info(y) {}
 };
 
-// Gets the llvm intrinsic ids on different platforms (NVPTX, AMDGPU)
-// corresponding to the give TargetIntrinsicID.
-struct TargetIntrinsics GetIntrinsic(TargetIntrinsicID intrin) {
-  switch (intrin) {
-    case TargetIntrinsicID::kShflDownF32: {
-      return {llvm::Intrinsic::nvvm_shfl_sync_down_f32,
-              llvm::Intrinsic::not_intrinsic};
+// Populates the function information for different platforms (NVPTX, AMDGPU)
+// corresponding to the given TargetFunctionID.
+// The call site provides arguments that is least restrictive to both targets.
+// For instance, if one target provides device function that takes 2 arguments
+// and the other provides arguments that takes 4 arguments, then the call
+// site provides four arguments. The following function  takes care of ignoring
+// the two superfluous argumenrs for the GPU device function that takes 2
+// arguments. In a similar vein, the following function also specifies
+// if conversions need to be generated for any argument type mismatches.
+// See check_signature  and use_bitcast fields in TargetFunctionInfo
+struct TargetInfo GetTargetInfo(TargetFunctionID function_id,
```

Comment:
It's pretty confusing to me that we have TargetFunctionID, which might be an intrinsic or a proper function, but then we have TargetIntrinsicInfo and TargetFunctionInfo.

I am also confused why we need a separate struct for intrinsics vs non-intrinsic device functions.  Intrinsics are a special case of functions, so maybe it would be simpler to put everything into one struct?